src/index.js

```diff
@@ -55,9 +61,8 @@ export default async (
               await cache.set(urlString, html);
             }
 
-            node.type = `html`;
-            node.value = html;
-            node.children = undefined;
+            // override the node with the html AST
+            paragraphNode.children[0] = htmlToHast(html);
```

Comment:
This is going in the right direction, but doesn’t work properly yet.

Let me try and walk through some of the background, which might help for understanding (but it might also be confusing, in which case the code suggestions are at the end 😅)

First: we’re in markdown, in an AST ([mdast](https://github.com/syntax-tree/mdast) to be specific). ASTs are nice! You can do stuff like what this plugin does!
What happened in these lines before though, was that we injected a string of HTML. That works. But we’re loosing ASTs!

Second: Gatsby, MDX, and many other projects, will use this markdown AST and later turn it into an HTML AST ([hast](https://github.com/syntax-tree/hast), to be specific, through remark-rehype). After that, they’ll do their own things: turning it into Vue `h()` calls, `React.createElement` calls, just serialized HTML, maybe print to a terminal, or whatnot.
Having an HTML AST in between is nice because the end targets don’t need to know about markdown: maybe someone writes in [orgmode](https://github.com/orgapp/orgajs), turns that into the HTML AST, and all those tools still work!

Alright: so we’re getting an HTML AST, and soon the markdown AST will be turned into an HTML AST. But: we’re in markdown, what do we do with an HTML AST?

It’s possible to define on a markdown node, how it should later be turned into an HTML node, [with a couple fields](https://github.com/syntax-tree/mdast-util-to-hast#notes)!

```suggestion
            const link = paragraphNode.children[0];
            const htmlElement = htmlToHast(html);
            
            link.data = {
              hName: htmlElement.tagName,
              hProperties: htmlElement.properties,
              hChildren: htmlElement.children
            };
```

(pseudo code, haven’t checked code style but it *should* roughly work)

Reply:
Thanks @wooorm! This didn't quite work. I think it's because the `link` variable there is pointing to a `node` with the type of `text`. What type should it be set to?

Reply:
oh `node` is [apparently the link](https://github.com/MichaelDeBoey/gatsby-remark-embedder/blob/57524b0be5760d30cef355a5aa84aa9c0fb1eb96/src/index.js#L37). So setting `node.data` should do the trick I think?

Reply:
But we want to totally replace the link node with an html element node.

Reply:
I believe the goal is somewhat different: to define what the node turns into when it is compiled to html/react/vue. Typically a link becomes an 'a', but by defining something else on node.data, it can become an Iframe or so!

Reply:
It might be an `iframe`, but in the case of twitter, it's actually a `<blockquote>`.

Take a look at the before/after in the README: https://github.com/MichaelDeBoey/gatsby-remark-embedder#twitter

Reply:
I may definitely be missing something and I don't know this project. But, before I dive into it myself: have you tried my suggestion out on 'node'? I'm under the assumption it'll work, because mdx, gatsby, next, etc will all move from markdown to html to whatever. Anyway, back at a computer tomorrow and i'll try it!

Reply:
Thanks for your help here @wooorm! I've pushed the suggested changes. Here's an example of the test output for these changes:

```
 FAIL  src/__tests__/transformers/SoundCloud.js
  ● Plugin can transform SoundCloud links

    expect(received).toMatchInlineSnapshot(snapshot)

    Snapshot name: `Plugin can transform SoundCloud links 1`

    - Snapshot  - 3
    + Received  + 3

    @@ -6,9 +6,9 @@

      <https://api.soundcloud.com/playlists/703823211>

      <https://w.soundcloud.com/player?url=https://soundcloud.com/clemenswenners/africa>

    - <iframe width="100%" height="300" scrolling="no" frameborder="no" src=https://w.soundcloud.com/player?url=https://soundcloud.com/clemenswenners/africa&color=ff5500&auto_play=false&hide_related=true&show_comments=true&show_user=true&show_reposts=false&show_teaser=false&visual=true></iframe>
    -
    - <iframe width="100%" height="300" scrolling="no" frameborder="no" src=https://w.soundcloud.com/player?url=http://soundcloud.com/clemenswenners/africa&color=ff5500&auto_play=false&hide_related=true&show_comments=true&show_user=true&show_reposts=false&show_teaser=false&visual=true></iframe>
    + <https://soundcloud.com/clemenswenners/africa>
    +
    + <http://soundcloud.com/clemenswenners/africa>
      ↵

      57 |   const processedAST = await plugin({ cache, markdownAST });
      58 | 
    > 59 |   expect(parseASTToMarkdown(processedAST)).toMatchInlineSnapshot(`
         |                                            ^
      60 |     "<https://not-a-soundcloud-url.com>
      61 | 
      62 |     <https://this-is-not-soundcloud.com>

      at Object.<anonymous> (src/__tests__/transformers/SoundCloud.js:59:44)
```

In this example instead of the `iframe` I get a link. I'm not sure what's going on...

Reply:
Ah. The tests serialize Markdown. They’re not testing whether it works in Gatsby or when going to HTML

Reply:
Basing this on https://github.com/MichaelDeBoey/gatsby-remark-embedder/pull/168 should show that nothing meaningful changes in this PR